Ready for review!

BrittanyBlake opened the pull request:
#Enumerables method

Created my own version of methods for the following methods:
 - #each
- #each_with_index
- #select
- #all?
- #any?
- #none?
- #count
- #inject
- #map

Thanks for reviewing my project! I look forward to your suggestions and comments.

BrittanyBlake committed:
Initial commit

BrittanyBlake committed:
Add my_each method

BrittanyBlake committed:
Add  my_each_with_index method

BrittanyBlake committed:
Add select method

BrittanyBlake committed:
fix rubocop errors

BrittanyBlake committed:
Add my_all? method

BrittanyBlake committed:
Add my_any? method

BrittanyBlake committed:
Add my_none? method

BrittanyBlake committed:
Fix rubocop errors

BrittanyBlake committed:
Add my_count

BrittanyBlake committed:
Add my_map method

BrittanyBlake committed:
Fix linter errors

BrittanyBlake committed:
Add my_inject

BrittanyBlake committed:
modify my_map

BrittanyBlake committed:
Add descriptive readme

BrittanyBlake committed:
Fix rubocop errors

mosaaleb commented on enumerable_methods.rb:
```diff
@@ -0,0 +1,132 @@
+  # end of my_each_with_index
+
+  # my_select
+  def my_select
```
- [ ] [`select`](https://apidock.com/ruby/Array/select) returns a new array containing all elements of array for which the given block returns true value.

Your current implementation doesn't return the `new_array`, it returns nil.

Also in `new_array << i` you are appending the index, not the element itself.

mosaaleb commented on enumerable_methods.rb:
```diff
@@ -0,0 +1,132 @@
+  # end of my_none?
+
+  # my_count
+  def my_count
```
- [ ] [`count`](https://rubyapi.org/2.7/o/array#method-i-count) works by expecting a block or an argument. In case of passing argument it will count the occurences of that argument.
```ruby
[1, 2, 3].count(2) # => 1
```

In case of passing a block, it will return the number of items yields true inside the block.

```ruby
[1, 2, 3].count(&:even?)
# or
[1, 2, 3].count { |x| x.even? }
# should give 1
```

You have implemented the block (not working right though -- `count` is increasing no matter what), but we have to implement the other case when just an argument is passed (`[1, 2, 3].my_count(2)`)

mosaaleb commented on enumerable_methods.rb:
```diff
@@ -0,0 +1,132 @@
+  # end of my_count
+
+  # start of my_map
+  def my_map(_proc = nil)
```
- [ ] [`map`](https://rubyapi.org/2.7/o/enumerable#method-i-map) Returns a new array with the results of running block once for every element in enum.

Your current implementation is returning the same array disregarding the passed block.

```ruby
[5, 2, 1].my_map { |x| x * 2 }
# should give [10, 2, 1]
# But the result is [5, 2, 1]
```

You're appending it to the array in `new_array << item` and ignoring the passed block. Let's append the yielded item from the block.

mosaaleb commented on enumerable_methods.rb:
```diff
@@ -0,0 +1,132 @@
+  # end of my_map
+
+  # start of my_inject
+  def my_inject(acc)
```
- [ ] [`inject`](https://rubyapi.org/2.7/o/enumerable#method-i-inject) takes initial value to start with and it's optional, for example:
```ruby
# without passing initial value to inject
[5, 2, 1].inject { |result, x| result + x }
# with passing initial value to inject
[5, 2, 1].inject(0) { |result, x| result + x }
```

In case of not passing initial value, ruby will select the first item of the array to be the initial value. The current implementation breaks if the initial value is not passed. Let's add a case if the initial value is not passed.

mosaaleb commented on enumerable_methods.rb:
```diff
@@ -0,0 +1,132 @@
+  def my_map(_proc = nil)
+    new_array = []
+    my_each do |item|
+      new_array << item
```
- [ ] `my_map` returns enum if no block was given.

mosaaleb reviewed (changes requested):
#### Good job @BrittanyBlake :clap:

Most requirements are handled, yet there are some requirements needs improvements before moving on.

Take a look at my comments below.


Please handle these requirements and submit a new code review.

##### Muhammad Ebeid

BrittanyBlake committed:
Modify my_select method

BrittanyBlake committed:
Modify my_count

BrittanyBlake committed:
modify my_count method

BrittanyBlake committed:
Modify my_map method

BrittanyBlake committed:
Modify my_inject method

BrittanyBlake committed:
Fix rubocop errors

BrittanyBlake commented:
#Enumerables method

As per the suggestions given, I have implemented the following changes:

- modified the #my_select method to append the element and return the new array.
- modified the #my_count method  to expect an argument or a block 
- fixed #my_count method so that it is no longer increasing no matter what 
- modified the #my_count method  to implement case when only an argument is passed
- modified #my_map method to return to enum if no block is given
-modified #my_map method to return a new array with the results 
- modified #my_inject method so that the initial value is optional and it no longer breaks if no initial value is passed

Thanks! I look forward to more suggestions to improve my code.

euqueme commented on enumerable_methods.rb:
```diff
@@ -0,0 +1,174 @@
+      return false unless yield item
+    end
+    true
+  end
```
* [ ] should have working `my_all?`  identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_all? 

- [ ] when no block or argument is given returns true only when none of the collection members are false or nil if the mentioned condition is not met it should return false

```ruby     
[nil, false, true, []].my_all? == [nil, false, true, []].all? => false
```
- Yours is giving LocalJumpError
![image](https://user-images.githubusercontent.com/52209067/78290610-31314e80-74e1-11ea-86f6-9775cc9339e6.png)

- Real all?
![image](https://user-images.githubusercontent.com/52209067/77688266-67148700-6f65-11ea-82e5-cbef16444817.png)

- [ ] when a class is passed as an argument returns true only if all of the collection is a member of such class, otherwise, it should return false

```ruby     
[1, 2.5, 'a', 9].my_all?(Integer) == [1, 2.5, 'a', 9].all?(Integer) => false
```
- Yours is retrieving a ArgumentError
![image](https://user-images.githubusercontent.com/52209067/78291095-0693c580-74e2-11ea-8d6d-0fb9d459a741.png)

- Real all?
![image](https://user-images.githubusercontent.com/52209067/78291125-17443b80-74e2-11ea-9f08-33b5a870ca79.png)

- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex

```ruby     
%w[dog door rod blade].my_all?(/d/) == %w[dog door rod blade].all?(/d/) => true
```
- Yours is retrieving a ArgumentError
![image](https://user-images.githubusercontent.com/52209067/78296760-8b371180-74eb-11ea-803c-7f8938ee7467.png)

- Real all?
![image](https://user-images.githubusercontent.com/52209067/77688741-31bc6900-6f66-11ea-90ec-b6a7cee7e6ca.png)

- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

```ruby     
[3,4,7,11].my_all?(3) == [3,4,7,11].all?(3)  => false 
```
- Yours is retrieving a ArgumentError
![image](https://user-images.githubusercontent.com/52209067/78296931-d94c1500-74eb-11ea-9016-7d47734a9b9f.png)

- Real all?
![image](https://user-images.githubusercontent.com/52209067/77689001-924ba600-6f66-11ea-9b72-210b3bb1d78b.png)

euqueme commented on enumerable_methods.rb:
```diff
@@ -0,0 +1,174 @@
+      return true if yield item
+    end
+    false
+  end
```
- [ ] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

 #### my_any? 

- [ ] when no block or argument is given returns true if at least one of the collection is not false or nil

```ruby     
[nil, false, true, []].my_any? == [nil, false, true, []].any? => true
```
- Yours returns an LocalJumpError
![image](https://user-images.githubusercontent.com/52209067/78297828-a9057600-74ed-11ea-9df0-a5ebb37b2421.png)

- Real any?
![image](https://user-images.githubusercontent.com/52209067/77689425-3f262300-6f67-11ea-88a9-2885ebab08d4.png)

- [ ] when a class is passed as an argument returns true if at least one of the collection is a member of such class

```ruby
[1, 2.5, 'a', 9].my_any?(Integer) == [1, 2.5, 'a', 9].any?(Integer) => true
```
- Yours is retrieving a ArgumentError
![image](https://user-images.githubusercontent.com/52209067/78297877-c89c9e80-74ed-11ea-997a-7965d4592a2b.png)

- Real any?
![image](https://user-images.githubusercontent.com/52209067/77689595-8f04ea00-6f67-11ea-9f51-4536c3f8049b.png)

- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex

```ruby
%w[dog door rod blade].my_any?(/z/) == %w[dog door rod blade].any?(/z/) => false
```
- Yours retrieves an ArgumentError
![image](https://user-images.githubusercontent.com/52209067/78297980-06012c00-74ee-11ea-829e-e092ad5abc04.png)

- Real any?
![image](https://user-images.githubusercontent.com/52209067/77689742-cf646800-6f67-11ea-9e3e-eec7a960fe38.png)

- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

```ruby
[3,4,7,11].my_any?(3) should be equal to this [3,4,7,11].any?(3) => true
```
- Yours retrieves an ArgumentError
![image](https://user-images.githubusercontent.com/52209067/78298026-23ce9100-74ee-11ea-82f1-41544b38c2c0.png)

- Real any?
![image](https://user-images.githubusercontent.com/52209067/77689883-10f51300-6f68-11ea-8a4c-319977bb3842.png)

euqueme commented on enumerable_methods.rb:
```diff
@@ -0,0 +1,174 @@
+      return false if yield item
+    end
+    true
+  end
```
- [ ] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_none?

- [ ] when no block or argument is given returns true only if none of the collection members is true
```ruby
[nil, false, true, []].my_none? == [nil, false, true, []].none? => false
```
- Yours is retrieving a LocalJumpError
![image](https://user-images.githubusercontent.com/52209067/78302837-d73c8300-74f8-11ea-8eae-c8527ae960bc.png)

- Real none?
![image](https://user-images.githubusercontent.com/52209067/77690365-cd4ed900-6f68-11ea-9933-f73b0091208d.png)

- [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
[1, 2.5, 'a', 9].my_none?(Integer) == [1, 2.5, 'a', 9].none?(Integer) => false
```
- Yours is retrieving an ArgumentError
![image](https://user-images.githubusercontent.com/52209067/78302892-f4715180-74f8-11ea-8729-c00f56382ad5.png)

- Real none?
![image](https://user-images.githubusercontent.com/52209067/77690539-130ba180-6f69-11ea-8bd0-6ca061ab05d2.png)

- [ ] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_none?(/z/) == %w[dog door rod blade].none?(/z/) => true
```
- Yours is retrieving an ArgumentError
![image](https://user-images.githubusercontent.com/52209067/78302979-1539a700-74f9-11ea-97a0-9e74668c153f.png)

- Real none?
![image](https://user-images.githubusercontent.com/52209067/77690722-65e55900-6f69-11ea-8881-9afe27d263dc.png)

- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
[3,4,7,11].my_none?(3) == [3,4,7,11].none?(3) => false
```
- Yours is retrieving an ArgumentError
![image](https://user-images.githubusercontent.com/52209067/78303068-369a9300-74f9-11ea-9bad-8e5591b48c7e.png)

- Real none?
![image](https://user-images.githubusercontent.com/52209067/77690915-a9d85e00-6f69-11ea-81b3-7adb07b281ca.png)

[OPTIONAL]

I think we could get the block of my_none? from my_any? we should only do something like !my_any? for my_none? to work. For example:
```ruby
  def my_none?(arg = nil, &block)
    !my_any?(arg = nil,&block)
  end
```
To make this work we have to use block.call(element) instead of yield(element) in my_any?

euqueme commented on enumerable_methods.rb:
```diff
@@ -0,0 +1,174 @@
+
+  # start of my_map
+
+  def my_map(_proc = nil)
```
[OPTIONAL]

- [ ] In the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend working on it.
Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```

euqueme commented on enumerable_methods.rb:
```diff
@@ -0,0 +1,174 @@
+  # end of my_map
+
+  # start of my_inject
+  def my_inject(initial = nil)
+    acc = if initial.nil?
+            0
+          else
+            initial
+          end
+
+    my_each do |i|
+      acc = yield(acc, i)
+    end
+    acc
+  end
```
- [ ] should have working `my_inject` **(covered by automated tests)** identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

####  my_inject

- [ ] combines all elements of enum by applying a binary operation, specified by a block
```ruby
(1..5).my_inject(4) { |prod, n| prod * n } == (1..5).inject(4) { |prod, n| prod * n } => 480
```
- Yours is retrieving a NameError
![image](https://user-images.githubusercontent.com/52209067/78304308-aad63600-74fb-11ea-80c5-8f4dff16b3b4.png)

- Real inject
![image](https://user-images.githubusercontent.com/52209067/77698409-2ffaa180-6f76-11ea-91d3-9778aa560411.png)

- [ ] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.
```ruby
[2, 5, 6].my_inject(:+) == [2, 5, 6].inject(:+) => 13
```
- Yours is retrieving a LocalJumpError
![image](https://user-images.githubusercontent.com/52209067/78304375-d35e3000-74fb-11ea-9679-051c26752069.png)

- Real inject
![image](https://user-images.githubusercontent.com/52209067/77698534-6a643e80-6f76-11ea-9d0e-ebb0f02da179.png)

euqueme commented on enumerable_methods.rb:
```diff
@@ -0,0 +1,174 @@
+  # end of my_inject
+
+  # multiply elements method
+  def multiply_els
+    my_inject(1) { |product, n| product * n }
+  end
```
[OPTIONAL]

- [ ] mutiply_els is not an Enumerable method so I strongly recommend to put it in a different module

euqueme commented on enumerable_methods.rb:
```diff
@@ -0,0 +1,174 @@
+
+#  p [1, 2, 3, 4, 5].select { |i|  i.even? }
+#  p [1, 2, 3, 4, 5].my_select { |i|  i.even? }
+#  p [1, 2, 3, 4, 5].select { |i|  i.odd? }
+#  p [1, 2, 3, 4, 5].my_select { |i|  i.odd? }
+
+# [2,2,2,1,5].all? {|i| p i.even?}
+# [2,2,2,1,5].my_all? {|i| p i.even?}
+
+# [2,2,2,1,5].any? {|i| p i.even?}
+# [2,2,2,1,5].my_any? {|i| p i.even?}
+
+#  [2,2,2,1,5].none? {|i| p i.even?}
+#  [2,2,2,1,5].my_none? {|i| p i.even?}
+
+#  p [2, 2, 2, 1, 5].count
+#  p [2, 2, 2, 1, 5].my_count
+#  p [1, 2, 3].count(&:even?)
+#  p [1, 2, 3].my_count(&:even?)
+#  p [1, 2, 3].count(2) # => 1
+#  p [1, 2, 3].my_count(2) # => 1
+
+# p [2, 2, 2, 1, 5].map
+# p [2, 2, 2, 1, 5].my_map
+# p [2, 2, 2, 1, 5].map { |i| i }
+# p [2, 2, 2, 1, 5].my_map { |i| i }
+# p [5, 2, 1].my_map { |x| x * 2 }
+# p [5, 2, 1].map { |x| x * 2 }
+
+#   p [3, 6, 10, 13].inject(:+)
+#   p [3, 6, 10, 13].my_inject {|sum, number| sum + number} #=> 32
+#   p [2,4,5].multiply_els
+# # without passing initial value to inject
+# # p [5, 2, 1].inject { |result, x| result + x }
+# # # with passing initial value to inject
+# # p [5, 2, 1].inject(0) { |result, x| result + x }
+# # without passing initial value to inject
+#  p [5, 2, 1].my_inject { |result, x| result + x } #=> 8
+# # with passing initial value to inject
+#  p [5, 2, 1].my_inject(0) { |result, x| result + x } #=> 8 works
```
* [ ] Comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

- Please delete these test cases or put them in a different file

euqueme reviewed (changes requested):
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @BrittanyBlake :handshake: 

### Thank you for doing the changes requested by the previous reviewer, :medal_sports:  your Enumerable methods are working better and better every time, :star:  there are some methods that needed changes but weren't mentioned by the previous reviewer so I'm very sorry for that :bowing_woman: there are also some optional changes so please address them in the comments I've put in your code bellow :point_down: 

#### General

* [ ] Files not related to the project should not be committed and added to .gitignore (e.g. `.DS_Store`) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/gitignore.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

- your `.DS_Store` file in your assets directory shouldn't be committed and added to the .gitignore

### That's it! :raised_hands: 

### Please :pray: make the changes :woman_technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any questions? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :traffic_light:

BrittanyBlake committed:
.DS_Store banished!

BrittanyBlake committed:
modify my_all? method

BrittanyBlake committed:
Fix module length errors, method length errors, and perceived complexity errors

BrittanyBlake committed:
modify my_any? method

BrittanyBlake committed:
Modify my_none? method

BrittanyBlake committed:
remove old my_none?(commented out) method

BrittanyBlake committed:
Fix rubocop errors

BrittanyBlake committed:
modify my_map method

BrittanyBlake committed:
remove multiply_els from module

BrittanyBlake committed:
modify my_inject method

BrittanyBlake committed:
fix rubocop errors

BrittanyBlake commented:
#Enumerables method

Thanks for the review @euqueme ! Your comments were extremely helpful. 

As per the comments given, I have implemented the following changes:

- removed ".DS_Store" files. 
-  modified the #my_all? method so it functions as requested by TSE
- modified the #my_any? method so it functions as requested by TSE
- modified the #my_none? method to get the block from my_any? as suggested by TSE
- implemented the optional suggestion and modified the #my_map method to take either a proc or a block 
- modified the #my_inject method so it functions as requested by TSE
- removed the commented out test cases as suggested by TSE

Thanks! I look forward to more suggestions to improve my code.

euqueme reviewed (changes requested):
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @BrittanyBlake :wave: 

### Thank you for doing all the changes I requested even the optional ones, you are a real team player :soccer: :medal_sports:  you've come a long way with this project so you should feel very proud of yourself :star: there are only 2 methods that need changes so please keep reading :point_down: 

#### my_any?
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

```ruby
%w[dog door rod blade].my_any?('cat') == %w[dog door rod blade].any?('cat') => false
```
- Yours retrieves a TypeError when checking Strings
![image](https://user-images.githubusercontent.com/52209067/78453561-cbafa000-764f-11ea-8f53-1c2f029dd4b2.png)

- Real any? checking Strings
![image](https://user-images.githubusercontent.com/52209067/78453573-d9652580-764f-11ea-9a10-57dabceeb6bc.png)

```ruby
[3,4,7,11].my_any?(3) should be equal to this [3,4,7,11].any?(3) => true
```

- Yours retrieves the wrong answer when checking Integers
![image](https://user-images.githubusercontent.com/52209067/78453594-fbf73e80-764f-11ea-8622-c32395158107.png)

- Real any? checking Integers
![image](https://user-images.githubusercontent.com/52209067/78453606-0fa2a500-7650-11ea-9d0c-65da54de67c1.png)

#### my_none?
 - I see you are getting my_none? from !my_any? so once you solve the changes for my_any? :point_up:  this enumerable should work properly.


### That's it! :raised_hands: 

### Please :pray: do the changes :woman_technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any questions? ask me on slack **@ Maru**

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

BrittanyBlake committed:
modify my_any? method

BrittanyBlake committed:
fix error

BrittanyBlake committed:
fix error

BrittanyBlake committed:
Fix linter errors

BrittanyBlake committed:
configure cops in rubocop file

BrittanyBlake committed:
fix errors

BrittanyBlake commented:
#Enumerables method

- As per the comments given, I have modified the #my_any? method. 

The linter errors are failing on the pull request but not when I run rubocop in my terminal:

- I have already configured the cops that it was telling me to add so I have no idea why it is continuing to fail the checks in this pull request. 

Thanks! I look forward to more suggestions to improve my code and fix the stickler error.

BrittanyBlake committed:
Trying to fix errors

BrittanyBlake committed:
Trying to fix errors

BrittanyBlake commented:
- Successfully fixed errors :)

majovanilla commented:
Hey @BrittanyBlake , 👋

## You did an *amazing*🌟 ✨ job on this project! 👊🏻  💯

## You passed all tests and your project is APPROVED!!! 😀

You are free to merge. Happy coding 🤓

--Maria Reyes
